pkg/controller/daemon/update.go

```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+	var result []*apps.ControllerRevision
+	selector, err := metav1.LabelSelectorAsSelector(ds.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	histories, err := dsc.historyLister.List(selector)
+	if err != nil {
+		return nil, err
+	}
+	for _, history := range histories {
+		// Skip history that doesn't belong to the DaemonSet
+		if controllerRef := controller.GetControllerOf(history); controllerRef == nil || controllerRef.UID != ds.UID {
+			continue
+		}
+		result = append(result, history)
+	}
+	return result, nil
+}
+
+// match check if ds template is semantically equal to the template stored in history
+func match(ds *extensions.DaemonSet, history *apps.ControllerRevision) (bool, error) {
+	template, err := decodeHistory(history)
+	return apiequality.Semantic.DeepEqual(&ds.Spec.Template, template), err
+}
+
+func decodeHistory(history *apps.ControllerRevision) (*v1.PodTemplateSpec, error) {
+	raw := history.Data.Raw
+	decoder := json.NewDecoder(bytes.NewBuffer(raw))
+	template := v1.PodTemplateSpec{}
+	err := decoder.Decode(&template)
+	return &template, err
+}
+
+func encodeTemplate(template *v1.PodTemplateSpec) ([]byte, error) {
+	buffer := new(bytes.Buffer)
+	encoder := json.NewEncoder(buffer)
+	err := encoder.Encode(template)
+	return buffer.Bytes(), err
+}
+
+// computeHash returns a hash value calculated from pod template and a uniquifier to avoid hash collision
+// TODO: this is copied from deployment controller util, will move to a shared library for all controllers
+func computeHash(template *v1.PodTemplateSpec, uniquifier *int64) uint32 {
+	podTemplateSpecHasher := fnv.New32a()
+	hashutil.DeepHashObject(podTemplateSpecHasher, *template)
+
+	// Add uniquifier in the hash if it exists.
+	if uniquifier != nil {
+		uniquifierBytes := make([]byte, 8)
+		binary.LittleEndian.PutUint64(uniquifierBytes, uint64(*uniquifier))
+		podTemplateSpecHasher.Write(uniquifierBytes)
+	}
+
+	return podTemplateSpecHasher.Sum32()
+}
+
+func (dsc *DaemonSetsController) snapshot(ds *extensions.DaemonSet, revision int64) (*apps.ControllerRevision, error) {
+	encodedTemplate, err := encodeTemplate(&ds.Spec.Template)
+	if err != nil {
+		return nil, err
+	}
+	hash := fmt.Sprint(computeHash(&ds.Spec.Template, ds.Status.Uniquifier))
+	name := ds.Name + "-" + hash
+	history := &apps.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Namespace:       ds.Namespace,
+			Labels:          labelsutil.CloneAndAddLabel(ds.Labels, extensions.DefaultDaemonSetUniqueLabelKey, hash),
+			Annotations:     ds.Annotations,
+			OwnerReferences: []metav1.OwnerReference{*newControllerRef(ds)},
+		},
+		Data:     runtime.RawExtension{Raw: encodedTemplate},
+		Revision: revision,
+	}
+
+	history, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Create(history)
+	if errors.IsAlreadyExists(err) {
+		// TODO: Is it okay to get from historyLister?
```

Comment:
AlreadyExists is from API server. If we saw this in cache (when trying to find current history of a DaemonSet) we wouldn't try to create it and instead just reuse the existing one.